File: langbook/extra/recursive/recursive.py
# -*- coding: utf-8 -*-

def is_integer(x):
    return isinstance(x, int)

def total(xs):
    result = 0
    for x in xs:
        if is_integer(x):
            result += x
        else:
            result += total(x)

    return result


def total2(xs):
    # 初期化
    result = 0
    i = 0

    # for文に相当
    while i < len(xs):
        x = xs[i]
        i += 1
        if is_integer(x):
            result += x
        else:
            result += total(x)

    return result


def total3(xs):
    stack = []
    is_function_call = True
    # gotoのない言語で「関数冒頭へのジャンプ」に相当することを実現するためにwhileを使う
    while True:
        if is_function_call: # 関数呼び出し時だけTrueになるフラグ
            # 初期化
            result = 0
            i = 0
            is_function_call = False

        # for文に相当
        while i < len(xs):
            x = xs[i]
            i += 1
            if is_integer(x):
                result += x
            else:
                ## 関数呼び出しに相当することをやる
                # 今の状態をスタックに積む
                stack.append((xs, i, result))
                # xsを変える(xを引数としてtotalを呼ぶことに相当)
                xs = x
                # ここで関数冒頭へgotoしたいところだがPythonにはgotoがないので
                # フラグを立ててbreakする
                is_function_call = True
                break

        # whileを抜けた際、ループが普通に終了した場合と、関数呼び出しのためにbreakしたのを区別する必要がある
        if is_function_call:
            # 関数呼び出しの場合、何もせずにcontinueすることでwhile冒頭へのジャンプを行う
            continue
        else:
            # 関数呼び出しでない場合

            # 元の関数のループの後に何らかの処理があるならばここに入るが、今回はたまたま
            # 何もせずにreturn resultで抜けているのでここに入る処理はない。

            # 関数を抜けた後の処理
            if not stack: # スタックが空なら
                return result # すべての計算が完了したのでこの関数から抜ける

            # スタックが積まれている場合、中断していた処理を再開する必要がある。
            # 積んでおいた値を取り出す
            old_xs, old_i, old_result = stack.pop()
            # 値を元に戻す
            xs = old_xs
            i = old_i
            result = old_result + result # result += total(x)で返り値を足していることに相当
            # この後、while冒頭に戻って、 is_function_call == False なので初期化処理を飛ばして
            # 2つ目のwhileに戻るので結果的に「再開」ができる。

    return result

print total([1, [2, 3], 4])
print total2([1, [2, 3], 4])
print total3([1, [2, 3], 4])
print total([1, [2, 3], 4, [1, [2, 3], 4]])
print total2([1, [2, 3], 4, [1, [2, 3], 4]])
print total3([1, [2, 3], 4, [1, [2, 3], 4]])

File: langbook/extra/recursive/recursive.cpp
#include <stack>
#include <tuple>
#include <vector>
#include <iostream>
#include <boost/any.hpp>
#include <boost/foreach.hpp>

using boost::any_cast;
typedef std::vector<boost::any> many;


bool is_integer(const boost::any& x){
  return (x.type() == typeid(int));
}


void show_many(const many& xs);

void show_any(const boost::any x){
  if(is_integer(x)){
    std::cout << any_cast<int>(x);
  }else{
    show_many(any_cast<many>(x));
  }
}


void show_many(const many& xs){
  std::cout << "[";
  for(int i = 0; i < xs.size(); i++){
    if(i > 0){
      std::cout << ", ";
    }
    show_any(xs[i]);
  }
  std::cout << "]";
}


int total(const many& xs){
  int result = 0;
  BOOST_FOREACH( const boost::any& x, xs ){
    #ifdef VERBOSE
    std::cout << "xs:";
    show_many(xs);
    std::cout << ", x:";
    show_any(x);
    std::cout << ", result:" << result << std::endl;
    #endif

    if(is_integer(x)){
      result += any_cast<int>(x);
    }else{
      result += total(any_cast<many>(x));
    }
  }
  return result;
}


int total2(const many& xs){
  int i;
  boost::any x;
  int result = 0;

  for(i = 0; i < xs.size(); i++){
    x = xs[i];

    #ifdef VERBOSE
    std::cout << "xs:";
    show_many(xs);
    std::cout << ", x:";
    show_any(x);
    std::cout << ", i:" << i << ", result:" << result << std::endl;
    #endif

    if(is_integer(x)){
      result += any_cast<int>(x);
    }else{
      result += total2(any_cast<many>(x));
    }
  }
  return result;
}

// ローカル変数を保存するためのスタック
typedef std::tuple<many, int, boost::any, int> frame_t;
std::stack<frame_t> stack;

// 返り値を保存するための変数
int function_result;

int total3(many& xs){
 ENTRYPOINT:
  int i;
  boost::any x;
  int result = 0;

  for(i = 0; i < xs.size(); i++){
    x = xs[i];

    #ifdef VERBOSE
    std::cout << "xs:";
    show_many(xs);
    std::cout << ", x:";
    show_any(x);
    std::cout << ", i:" << i << ", result:" << result << std::endl;
    #endif

    if(is_integer(x)){
      result += any_cast<int>(x);
    }else{
      // result += total(any_cast<many>(x))相当のことを実現するために
      // 1: 現在のローカル変数をスタックに保存する
      stack.push(make_tuple(xs, i, x, result));
      // 2: 引数xsを書き換える
      xs = any_cast<many>(x);
      // 3: 関数冒頭へジャンプ
      goto ENTRYPOINT;

    RETURNPOINT: // 6: 呼び出された関数からreturnするとここに戻ってくる
      // 7: スタックに保存しておいた値を復元する
      frame_t f = stack.top();
      stack.pop();
      xs = std::get<0>(f);
      i = std::get<1>(f);
      x = std::get<2>(f);
      result = std::get<3>(f);

      // 8: 関数の返り値を使う
      result += function_result;
    }
  }

  // ループが終わったのでreturn result;に相当することをやる
  if(!stack.empty()){
    // スタックが空でないなら
    // 4: 返り値を決められた場所に保存
    function_result = result;
    // 5: 関数呼び出し直後(上記6)に戻る
    goto RETURNPOINT;
  }
  // スタックが空の時は自前管理の呼び出しではないので本物のreturnをする
  return result;
}


int main(){
  many xs;
  xs.push_back((boost::any) 1);

  many tmp;
  tmp.push_back((boost::any) 2);
  tmp.push_back((boost::any) 3);
  xs.push_back(tmp);

  xs.push_back((boost::any) 4);

  std::cout << total(xs) << std::endl;
  std::cout << total2(xs) << std::endl;
  std::cout << total3(xs) << std::endl;
}
